content/blog/2022-07-28-switching-to-dvc-from-jupyter-vscode.md

```diff
@@ -0,0 +1,283 @@
+predictions = predictions_by_class[:, 1]
+
+print(predictions)
+
+precision, recall, prc_thresholds = metrics.precision_recall_curve(labels_pred, predictions, pos_label=1)
+
+fpr, tpr, roc_thresholds = metrics.roc_curve(labels_pred, predictions, pos_label=1)
+
+avg_prec = metrics.average_precision_score(labels_pred, predictions)
+roc_auc = metrics.roc_auc_score(labels_pred, predictions)
+
+nth_point = math.ceil(len(prc_thresholds) / 1000)
+prc_points = list(zip(precision, recall, prc_thresholds))[::nth_point]
+
+with open(scores_file, "w") as fd:
+    json.dump({"avg_prec": avg_prec, "roc_auc": roc_auc}, fd, indent=4)
+
+with open(prc_file, "w") as fd:
+    json.dump(
+        {
+            "prc": [
+                {"precision": p, "recall": r, "threshold": t}
+                for p, r, t in prc_points
+            ]
+        },
+        fd,
+        indent=4,
+    )
+
+with open(roc_file, "w") as fd:
+    json.dump(
+        {
+            "roc": [
+                {"fpr": fp, "tpr": tp, "threshold": t}
+                for fp, tp, t in zip(fpr, tpr, roc_thresholds)
+            ]
+        },
+        fd,
+        indent=4,
+    )
+```
+
+Now you have all of the steps that you executed in your Jupyter notebook in a
+couple of files that you can easily edit and track across all of your
+experiments. This is a great time to commit these changes to your Git repo with
+the following commands:
+
+```cli
+$ git add train.py evaluate.py
+$ git commit -m "converted notebook to Python"
+```
+
+## Create the DVC pipeline
+
+Now we can create a DVC pipeline that executes these scripts to record the code,
+data, and metrics for each of your experiments. If you look in the project's
+`dvc.yaml`, you'll see the stages we execute on an experiment run.
+
+```yaml
+stages:
+  train:
+    cmd: python src/train.py ./data/ ./models/model.pkl
+    deps:
+      - ./data/train.pkl
+      - ./src/train.py
+    params:
+      - train.seed
+      - train.n_est
+      - train.min_split
+    outs:
+      - ./models/model.pkl
+  evaluate:
+    cmd:
+      python ./src/evaluate.py ./models/model.pkl ./data scores.json prc.json
+      roc.json
+    deps:
+      - ./data
+      - ./models/model.pkl
+      - ./src/evaluate.py
+    metrics:
+      - scores.json:
+          cache: false
+```
+
+The `stages` tell DVC which steps you want to execute and what should happen in
+each step. Usually, you'll execute a script or a command in each stage that may
+link to the next stage in the pipeline via the `outs`. We only have 2 stages in
+this pipeline: a `train` stage that handles the model training and outputs the
+model and an `evaluate` stage that takes the model and stores some metrics about
+it.
+
+Each of these stages has a `cmd` that executes the Python scripts we wrote with
+the required arguments. They both have defined dependencies in `deps` that let
+DVC know what needs to be available for a stage to execute before it starts
```

Comment:
To make it even more explicit, maybe add something along the lines here of

"As you can see, for example, the training stage is listed as a requirement for the evaluation stage. This ensures that the latter will only start once the first has been completed."

Reply:
And maybe as an admonition somewhere saying that when using `dvc exp run` only the stages downstream from your changes are triggered?